### DISCUSSION TITLE
A couple of formatting questions on constructor initialisers and aggregates

### DISCUSSION THREAD

**Discussion by jmccabe** [uncrustify/uncrustify#4000] [Q&A]
Hi,

Hope this isn't a FAQ, but I'm new to uncrustify (0.76) and finding the number of settings and dependencies slightly overwhelming at the moment. Having fiddled around with some of them, there are two things, at the moment, that stick out that I'm finding difficult to work out the settings for.

Firstly, in constructor initializer lists, we're using braced initialization, so may have something like:

```c++
ClassName::ClassName(int value1, double value2)
    : m_value1 { value1 },
      m_value2 { value2 }
{
```
In pretty much all the places where the space settings related to braces are configured, I've set them to put spaces in such that I'd expect the layout to be as above, but it's being changed to:

```c++
ClassName::ClassName(int value1, double value2)
    : m_value1{value1},
      m_value2{value2}
{
```
does anyone know if there's something obvious I may have missed?

The second areas is aggregates. Let's say I have (originally) something like:

```c++
constexpr std::pair<uint16_t, uint8_t>
    k_myTable[] { { 1u, 2u }, { 3u, 4u }, { 5u, 6u },
                  { 7u, 8u }, { 9u, 0u }, { 1u, 2u },
                  { 3u, 4u }, { 5u, 6u }, { 7u, 8u } };
```
This is being converted to:
```c++
constexpr std::pair<uint16_t, uint8_t>
k_myTable[] { { 1u, 2u }, { 3u, 4u }, { 5u, 6u },
    { 7u, 8u }, { 9u, 0u }, { 1u, 2u },
    { 3u, 4u }, { 5u, 6u }, { 7u, 8u } };
```
I'm not sure which settings might let me revert to the original version? Presumably there may be an 'ignore' one (or two), but I'd quite like that indentation to be actually applied where the layout doesn't match like that, e.g. if someone wrote the code like that latter one, it could be converted to the former.

Any help would be very gratefully appreciated.

**Comment by guy-maurel** [comment]
You don't need to try to guess which space-options are available for your source.
You need to use the CLI-option --tracking_space, such as:
`uncrustify -c jmccabe.cfg -f jmccabe.cpp --tracking_space jmccabe.html`
Have a look to the output file jmccabe.html with a browser.
The paces where a option is working are shown under the red M.

The second areas is very difficult. uncrustify doesn't have yet any algorithm to detect the structure of aggregates, which might be very complicated. It is a known problem.

**Reply by jmccabe** [reply]
Thank you for your reply. Could you show me an example of the output, perhaps? I've tried this on some of my real files and don't see any sign of a red M (BTW - this is on Ubuntu 20.04, with uncrustify 0.76 built from sources yesterday). I do see red Ms with I use tracking_nl though.

**Comment by guy-maurel** [comment]
tracking_nl is not yet useable, it is a project

**Reply by jmccabe** [reply]
Fair enough, but the output from tracking_nl, so far, has been more useful than tracking_space!
